File: typescript/src/sorters/greedy.ts
import {zip, combinations, md5} from '../utils'
import {Scalar, IncompletedType} from '../types'


interface sortArgsType {
  row: any,
  parents: Map<number, Scalar>,
  length: number,
  seed: any,
}

const ascendant = (a: number, b: number) => a > b ? 1 : -1

const getNumRemovablePairs = (
  indexes: number[],
  incompletedKeys: Set<string>, length: number
) => {
  let num = 0
  for (let vs of combinations(indexes, length)) {
    const key = vs.sort(ascendant).toString()
    if (incompletedKeys.has(key)) {
      num ++
    }
  }
  return num
}


export default function* (
  incompleted: IncompletedType,
  sortArgs: sortArgsType
) {
  const {row, parents, length} = sortArgs
  const seed = sortArgs.seed || ''
  const comparer = (a: [string, number[]], b: [string, number[]]) => {
    return md5(`${a[0]} ${seed}`) > md5(`${b[0]} ${seed}`) ? 1 : -1
  }

  while (true) {
    let maxNumPairs: number | null = null
    let efficientPair: number[] | null = null
    for (let [_, pair] of [... incompleted].sort(comparer)) {
      const keys = pair.map(p => parents.get(p) || 0)
      const candidate = zip(keys, pair)
      if (!row.storable(candidate)) {
        continue
      }
      const incompletedKeys = new Set(incompleted.keys())
      const numPairs = getNumRemovablePairs([... row.values(), ...pair], incompletedKeys, length)
      if (maxNumPairs === null || maxNumPairs < numPairs) {
        maxNumPairs = numPairs
        efficientPair = pair
      }
    }
    if (efficientPair === null) {
      break
    }
    yield efficientPair
  }
}

File: python/covertable/sorters/greedy.py
"""Greedy sorter
"""

import hashlib
import random
from collections import defaultdict
from itertools import product, combinations, permutations


def get_num_removable_pairs(indexes, incompleted, length):
    removing_keys = {
        tuple(sorted(vs))
        for vs in combinations(indexes, length)
    }
    return len(removing_keys & incompleted)


def sort(incompleted, row, parents, length, seed="", *args, **kwargs):
    def comparer(v):
        pair = ",".join(map(str, v))
        return hashlib.md5("{} {}".format(pair, seed).encode("utf-8")).hexdigest()

    while True:
        max_num_pairs = None
        efficient_pair = None
        for pair in sorted(incompleted, key=comparer):
            if not row:
                efficient_pair = pair
                break
            candidate = [(parents[p], p) for p in pair]
            if not row.storable(candidate):
                continue

            num_pairs = get_num_removable_pairs([*row.values(), *pair], incompleted, length)
            if max_num_pairs is None or max_num_pairs < num_pairs:
                max_num_pairs = num_pairs
                efficient_pair = pair
        if not efficient_pair:
            break

        yield efficient_pair
